docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,615 @@
+    var oldRoot = await document.GetSyntaxRootAsync(cancellationToken);
+    var newRoot = oldRoot.ReplaceNode(localDeclaration, formattedLocal);
+
+    // Return document with transformed tree.
+    return document.WithSyntaxRoot(newRoot);
+}
+```
+
+Your code fix is ready to try. Run your tests, and the new test that verifies the fix should pass. Then, try it interactively. Press F5 to run the Analyzer project in a second instance of Visual Studio. In the second Visual Studio instance, create a new C# Console Application project and, like before, add a few local variable declarations initialized with to constant values in the Main method.
+
+```csharp
+static void Main(string[] args)
+{
+    int i = 1;
+    int j = 2;
+    int k = i + j;
+}
+```
+
+You'll see that they are reported as warnings and "light bulb" suggestions appear next to them when the editor caret is on the same line. Move the editor caret to one of the squiggly underlines and press Ctrl+. to display the suggestion. Notice that a preview window appears next to the suggestion menu showing what the code will look like after the Code Fix is invoked.
+
+### Improve the quality
+
+Your analyzer works, but some C# expressions that aren't handled correctly:
+
+* The Diagnostic Analyzer's `AnalyzeNode` method does not check to see if the constant value is convertible to the variable type. So, the current implementation will happily convert an incorrect declaration such as int i = "abc"' to a local constant. This test will check that condition:
+
+[!code-csharp[Mismatched types don't raise diagnostics](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#AssignStringToInt "When the variable type and the constant type don't match, there's no diagnostic")]
+
+* Reference types are not handled properly. The only constant value allowed for a reference type is `null`, except in this case of <xref:System.String?displayPropert=nameWIthType>, which allows string literals. In other words, `const string s = "abc"` is legal, but `const object s = "abc"` is not. These two tests verify that condition:
+
+[!code-csharp[Reference types don't raise diagnostics](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#NoReferenceTypes "When the variable type is a reference type other than string, there's no diagnostic")]
+
+* If a variable is declared with the `var` keyword, the Code Fix does the wrong thing and generates a `const var` declaration, which is not supported by the C# language. To fix this bug, the code fix must replace the `var` keyword with the inferred type's name.
+Fortunately, all of the above bugs can be addressed using the same techniques that you just learned.
+
+[!code-csharp[Var declarations require types when constant](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#ReplaceVarDeclarationWithConst "Var must be replaced with the type when made const")]
```

Comment:
I think this snippet and the two above won't be formatted properly as part of the list and should be indented.